src/persist-client/src/internal/state_diff.rs

```diff
@@ -629,6 +653,120 @@ fn sniff_compaction<'a, T: Timestamp + Lattice>(
+///
+/// - [0,1),[1,3) (0,1 is empty)
+///
+/// This can only happen when the batch needing to be split is empty, so error
+/// out if it isn't because that means something unexpected is going on.
+///
+/// TODO: This implementation is certainly not correct if T is actually only
+/// partially ordered.
+fn apply_compaction_lenient<'a, T: Timestamp + Lattice>(
+    metrics: &Metrics,
+    mut trace: Vec<HollowBatch<T>>,
+    replacement: &'a HollowBatch<T>,
+) -> Result<Vec<HollowBatch<T>>, String> {
+    let mut removed_batches = Vec::new();
+    trace.retain(|b| {
+        let before_output = PartialOrder::less_equal(b.desc.upper(), replacement.desc.lower());
+        let after_output = PartialOrder::less_equal(replacement.desc.upper(), b.desc.lower());
+        let overlaps_output = !(before_output || after_output);
+        if overlaps_output {
+            removed_batches.push(b.clone());
+        }
+        !overlaps_output
+    });
+
+    {
+        let first_removed_batch = match removed_batches.first() {
+            Some(x) => x,
+            None => return Err("output didn't overlap any batches".into()),
+        };
+        if PartialOrder::less_than(first_removed_batch.desc.lower(), replacement.desc.lower()) {
+            if first_removed_batch.len > 0 {
+                return Err(format!(
+                    "removed batch was unexpectedly non-empty: {:?}",
+                    first_removed_batch
+                ));
+            }
+            let desc = Description::new(
+                first_removed_batch.desc.lower().clone(),
+                replacement.desc.lower().clone(),
+                first_removed_batch.desc.since().clone(),
+            );
+            trace.push(HollowBatch {
+                desc,
+                parts: Vec::new(),
+                len: 0,
+                runs: Vec::new(),
+            });
+            metrics.state.apply_spine_fast_path_lenient_adjustment.inc();
+        }
+    }
+
+    {
+        let last_removed_batch = match removed_batches.last() {
+            Some(x) => x,
+            None => return Err("output didn't overlap any batches".into()),
+        };
+        if PartialOrder::less_than(replacement.desc.upper(), last_removed_batch.desc.upper()) {
+            if last_removed_batch.len > 0 {
+                return Err(format!(
+                    "removed batch was unexpectedly non-empty: {:?}",
+                    last_removed_batch
+                ));
+            }
+            let desc = Description::new(
+                replacement.desc.upper().clone(),
+                last_removed_batch.desc.upper().clone(),
+                last_removed_batch.desc.since().clone(),
+            );
+            trace.push(HollowBatch {
+                desc,
+                parts: Vec::new(),
+                len: 0,
+                runs: Vec::new(),
+            });
+            metrics.state.apply_spine_fast_path_lenient_adjustment.inc();
```

Comment:
big fan of extending the spine metaphor to including spinal adjustments 😄

Reply:
realized when i looked at this again that this is actually a "slow" path (it reconstructs the spine), so renamed this metric